File: tests/Test_SWCharEst.f90
!-------------------------------------------------------------------------------------------------------------
! file: Test_SWCharEst.f90
! Test of module SWCharEst.
! build:
!    gfortran -g -o Test_SWCharEst ../src/SWCharEst.f90 Test_SWCharEst.f90
! run:
!    ./Test_SWCharEst
!-------------------------------------------------------------------------------------------------------------

subroutine DisplaySWCharEst ( name, values )

    implicit none
    character (len = *) :: name
    real :: values(1:4)

    print*, "  ", name, ": WP, FC, thetaS, Ks = ", &
	    values(1), ", ",  &
	    values(2), ", ",  &
	    values(3), ", ",  &
	    values(4)

end subroutine DisplaySWCharEst

! Returns true if fabs(a-b) / a <= threshold (a != 0)
! or if fabs(a-b) / b <= threshold (b != 0)
! or true if a = b = 0.
function AreClose ( a, b, threshold ) result( isClose )

    implicit none
    real :: a, b, threshold
    logical :: isClose

    isClose = .true.
    if ( 1.0 + (a - b) - 1.0 .ne. 0.0 ) then
	    if ( a .ne. 0.0 ) then
		isClose = ( abs ((a - b) / a) .le. threshold )
	    else if ( b .ne. 0.0 ) then
		isClose = ( abs ((a - b) / b) .le. threshold )
	    end if
    end if

end function AreClose

subroutine Compare ( expected, results )

    implicit none
    real :: expected(1:4), results(1:4)
    logical :: AreClose, passed

    passed = AreClose( expected(1), results(1), 1.0e-4 ) .and. &
	     AreClose( expected(2), results(2), 1.0e-4 ) .and. &
	     AreClose( expected(3), results(3), 1.0e-4 ) .and. &
	     AreClose( expected(4), results(4), 1.0e-3 )

    if ( passed ) then
	print*, "  passed"
    else
	print*, "  failed"
    endif

end subroutine Compare


program Test_SWCharEst

    use SWCharEst

    call Usage()
    call Test1()
    call Test2()

    contains

	subroutine Test1 ()
	    ! Example - sand:

	    implicit none
	    ! sand fraction, clay fraction, organic matter wt %
	    real :: soilTexture(1:3) = (/ 0.85, 0.04, 2.08 /)
	    !                          WP      FC       thetaS  Ks
	    real :: expected(1:4) = (/ 0.0400, 0.09785, 0.4545, 0.003096 /)
	    real :: results(1:4)

	    print*, "Test: SWCharEst( 0.85, 0.04, 2.08 )"
	    results = Get( soilTexture(1), soilTexture(2), soilTexture(3) );
	    call DisplaySWCharEst( "expected", expected );
	    call DisplaySWCharEst( "results", results );
	    call Compare( expected, results )

	end subroutine Test1

	subroutine Test2 ()
	    ! Example - silt loam:

	    implicit none
	    ! sand fraction, clay fraction, organic matter wt %
	    real :: soilTexture(1:3) = (/ 0.15, 0.18, 3.05 /)
	    !                          WP      FC       thetaS  Ks
	    real :: expected(1:4) = (/ 0.1286, 0.33148, 0.5050, 0.000433 /)
	    real :: results(1:4)

	    print*, "Test: SWCharEst( 0.15, 0.18, 3.05 )"
	    results = Get( soilTexture(1), soilTexture(2), soilTexture(3) );
	    call DisplaySWCharEst( "expected", expected );
	    call DisplaySWCharEst( "results ", results );
	    call Compare( expected, results )

	end subroutine Test2

end program

File: tests/Test_SWCharEst.cpp
// file:	Test_SWCharEst.cpp
// 		Test of class teh::SWCharEst
// build:
//	g++ -std=c++11 -g -Wall -I../src -o Test_SWCharEst Test_SWCharEst.cpp ../src/SWCharEst.cpp
// run:
//	./Test_SWCharEst

#include <iostream>
using std::cout;
using std::endl;
#include <vector>
#include <cmath>
#include "SWCharEst.h"
using teh::SWCharEst;

//	Returns true if fabs(a-b) / a <= threshold (a != 0)
//	or if fabs(a-b) / b <= threshold (b != 0)
//	or true if a = b = 0.
template
<
    typename T	///< floating-point type
>
inline bool AreClose (
	T const a, T const b, T const threshold )
{
	bool result = true;
	if ( 1.0f + (a - b) - 1.0f != 0.0f )
	{
		if ( a != 0.0f )
		    result = (std::fabs ((a - b) / a) <= threshold);
		else if ( b != 0.0f )
		    result = (std::fabs ((a - b) / b) <= threshold);
	}
	return result;
}

void DisplaySWCharEst (
    char const * const name,
    std::vector<float> const & values )
{
    cout << "  " << name << ": WP, FC, thetaS, Ks = "
	 << values[0] << ", "
	 << values[1] << ", "
	 << values[2] << ", "
	 << values[3] << endl;
}

void Compare (
    std::vector<float> const & expected,
    std::vector<float> const & results)
{
    bool passed = AreClose( expected[0], results[0], 1.0e-4f ) &&
		  AreClose( expected[1], results[1], 1.0e-4f ) &&
		  AreClose( expected[2], results[2], 1.0e-4f ) &&
		  AreClose( expected[3], results[3], 1.0e-3f );
    if ( passed )
	cout << "  passed" << endl;
    else
	cout << "  failed" << endl;
}

void Test1 ()
{
    cout << "Test: SWCharEst( 0.85, 0.04, 2.08 )" << endl;

    // sand fraction, clay fraction, organic matter wt %
    std::vector<float> const soilTexture = { 0.85, 0.04, 2.08 };

    //                                    WP      FC       thetaS  Ks
    std::vector<float> const expected = { 0.0400, 0.09785, 0.4545, 0.003096 };

    SWCharEst swc;
    std::vector<float> const results = swc.Get( soilTexture );
    DisplaySWCharEst( "expected", expected );
    DisplaySWCharEst( "results ", results );
    Compare( expected, results );
}

void Test2 ()
{
    cout << "Test: SWCharEst( 0.15, 0.18, 3.05 )" << endl;

    // sand fraction, clay fraction, organic matter wt %
    std::vector<float> const soilTexture = { 0.15, 0.18, 3.05 };

    //                                    WP      FC       thetaS  Ks
    std::vector<float> const expected = { 0.1286, 0.33148, 0.5050, 0.000433 };

    SWCharEst swc;
    std::vector<float> const results = swc.Get( soilTexture );
    DisplaySWCharEst( "expected", expected );
    DisplaySWCharEst( "results ", results );
    Compare( expected, results );
}

int main ()
{
    SWCharEst::Usage();
    Test1();
    Test2();
    return 0;
}
